### ISSUE TITLE
Module cannot be used as a type

### ISSUE THREAD

**Issue by JulianOrteil** [microsoft/pyright#2704]
**Describe the bug**
Related to https://github.com/microsoft/pylance-release/issues/1956

This warning is only displayed when used in a typing object like `List` or `Union`. Standalone (i.e. `c: a = None` where `a` is a module) displays no such warning by pylance (or pyright). As I am sure you can also gather, intellisense refuses to show any objects in the module causing the warning (i.e. the `c` variable below will not show `hello` as an option to use.

**To Reproduce**

**a.py**
```
hello: str = "world"
```

**b.py**
```
from typing import Union
from . import a

c: Union[a, str] = ""
```

**Expected behavior**
No warning.

**Screenshots or Code**
MRE above.

**VS Code extension or command-line**
Extension through Pylance

**Additional context**
Using Pylance v2021.12.2

**Comment by erictraut** [comment]
Pyright is correct to generate an error in this case.

It should also emit an error in the standalone case as well. In other words `c: a = None` should generate an error because a module is not valid to use in a type annotation.

**Comment by erictraut** [comment]
I've added an additional check so the standalone case also generates an error. This will be included in the next release.

**Comment by JulianOrteil** [comment]
I assume this is defined in one of the PEPs but I feel like removing this entirely is wrong.

It isn't really any different than saying `from . import a as a_better_name` IMO. Plus you can't work around this with certain frameworks like Django that require a settings.py module.

**Comment by erictraut** [comment]
I don't follow you. The statement `from . import a as a_better_name` creates a local variable called `a_better_name` which is equal to the imported module. A variable is very different from a type annotation.

A module within a type annotation is nonsensical. I don't know what it would even mean to be type compatible with a module.

For example, consider the following:
```python
import math
x: math = 1
```

What would a type checker do to determine whether `1` is type compatible with the module `math`?

**Comment by JulianOrteil** [comment]
Don't worry about my frustration eric, but thank you for your explanation.

I'm just attempting to use type annotations in a way they weren't intended to be used. I failed to see that PEP 484 only lists specific objects as acceptable type hints and modules is not one of them, so my apologies.

For context, the issue I'm trying to work around is the Django settings module requirement. Django provides a settings object (`from django.conf import settings`) that combines the aggregated local settings and global default settings into a single interface. It is improper to import the settings module directly and developers are instead directed to use the settings object in the import above. The problem though is Pylance (or pyright) cannot infer the variables set in the module from the object above, so they are never presented as an option for intellisense (I have hundreds of settings in this module due to the size of this Django site).

As an example:
```python
from typing import Union
from django.conf import settings as dj_settings
from . import settings as site_settings

settings: Union[dj_settings, site_settings] = dj_settings

...

if settings.DEBUG:
   ...
```

I was just *this* close to getting it working, only to find out that I was unintentionally exploiting a bug haha. Anyways, thank you for looking into this.

Jules

**Comment by JelleZijlstra** [comment]
One way that could work (not sure whether Pyright supports it) is to define a Protocol with all the fields present on the settings module and use that, because modules should be compatible with Protocols. But with hundreds of settings, that may not be very ergonomic. My typechecker, [pyanalyze](https://github.com/quora/pyanalyze), would support writing `Literal[dj_settings]`, but that wouldn't help you with autocompletion.

**Comment by erictraut** [comment]
That's an interesting suggestion, but I agree that maintaining a Protocol definition with all of those settings entries might be cumbersome. Pyright does support type validation of modules based on a Protocol definition. See [PEP 544](https://www.python.org/dev/peps/pep-0544/#modules-as-implementations-of-protocols) for details.

**Comment by JulianOrteil** [comment]
I will look into that option, thank you for the suggestions!

**Comment by erictraut** [comment]
This is included in pyright 1.1.196, which I just published. It will also be included in the next release of pylance.
